File: mcs/class/System.Drawing/System.Drawing.Drawing2D/Matrix.cs
//
// System.Drawing.Drawing2D.Matrix.cs
//
// Authors:
//   Stefan Maierhofer <sm@cg.tuwien.ac.at>
//   Dennis Hayes (dennish@Raytek.com)
//   Duncan Mak (duncan@ximian.com)
//   Ravindra (rkumar@novell.com)
//
// (C) Ximian, Inc.  http://www.ximian.com
// (C) Novell, Inc.  http://www.novell.com
//

using System;
using System.Drawing;
using System.Runtime.InteropServices;

namespace System.Drawing.Drawing2D
{
        public sealed class Matrix : MarshalByRefObject, IDisposable
        {
                internal IntPtr nativeMatrix;
                
                // constructors
                internal Matrix (IntPtr ptr)
                {
                        nativeMatrix = ptr;
                }
                
                public Matrix ()
                {
			Status status = GDIPlus.GdipCreateMatrix (out nativeMatrix);
			GDIPlus.CheckStatus (status);
                }
        
                public Matrix (Rectangle rect , Point[] plgpts)
                {
			Status status = GDIPlus.GdipCreateMatrix3I (rect, plgpts, out nativeMatrix);
			GDIPlus.CheckStatus (status);
                }
        
                public Matrix (RectangleF rect , PointF[] pa)
                {
			Status status = GDIPlus.GdipCreateMatrix3 (rect, pa, out nativeMatrix);
			GDIPlus.CheckStatus (status);
                }

                public Matrix (float m11, float m12, float m21, float m22, float dx, float dy)
                {
			Status status = GDIPlus.GdipCreateMatrix2 (m11, m12, m21, m22, dx, dy, out nativeMatrix);
			GDIPlus.CheckStatus (status);
                }
        
                // properties
                public float[] Elements {
                        get {
                                IntPtr tmp = Marshal.AllocHGlobal (Marshal.SizeOf (typeof (float)) * 6);
                                float [] retval = new float [6];

				Status status = GDIPlus.GdipGetMatrixElements (nativeMatrix, tmp);
				GDIPlus.CheckStatus (status);

                                Marshal.Copy (tmp, retval, 0, 6);

                                Marshal.FreeHGlobal (tmp);
                                return retval;
                        }
                }
        
                public bool IsIdentity {
                        get {
                                bool retval;
				Status status = GDIPlus.GdipIsMatrixIdentity (nativeMatrix, out retval);
				GDIPlus.CheckStatus (status);
                                return retval;
                        }
                }
        
                public bool IsInvertible {
                        get {
                                bool retval;
				Status status = GDIPlus.GdipIsMatrixInvertible (nativeMatrix, out retval);
				GDIPlus.CheckStatus (status);
                                return retval;
                        }
                }
        
                public float OffsetX {
                        get {
                                return this.Elements [4];
                        }
                }
        
                public float OffsetY {
                        get {
                                return this.Elements [5];
                        }
                }

                public Matrix Clone()
                {
                        IntPtr retval;
                        Status status = GDIPlus.GdipCloneMatrix (nativeMatrix, out retval);
			GDIPlus.CheckStatus (status);
                        return new Matrix (retval);
                }
                
        
                public void Dispose ()
                {
			Status status = GDIPlus.GdipDeleteMatrix (nativeMatrix);
			GDIPlus.CheckStatus (status);
                }                       
        
                public override bool Equals (object obj)
                {
                        Matrix m = obj as Matrix;

                        if (m != null) {
                                bool retval;
				Status status = GDIPlus.GdipIsMatrixEqual (nativeMatrix, m.nativeMatrix, out retval);
				GDIPlus.CheckStatus (status);
                                return retval;

                        } else
                                return false;
                }
        
                ~Matrix()
                {
                        Dispose ();
                }
                
                public override int GetHashCode ()
                {
                        return base.GetHashCode ();
                }
        
                public void Invert ()
                {
			Status status = GDIPlus.GdipInvertMatrix (nativeMatrix);
			GDIPlus.CheckStatus (status);
                }
        
                public void Multiply (Matrix matrix)
                {
                        Multiply (matrix, MatrixOrder.Prepend);
                }
        
                public void Multiply (Matrix matrix, MatrixOrder order)
                {
			Status status = GDIPlus.GdipMultiplyMatrix (nativeMatrix, matrix.nativeMatrix, order);
			GDIPlus.CheckStatus (status);
                }
        
                public void Reset()
                {
			Status status = GDIPlus.GdipSetMatrixElements (nativeMatrix, 1, 0, 0, 1, 0, 0);
			GDIPlus.CheckStatus (status);
                }
        
                public void Rotate (float angle)
                {
                        Rotate (angle, MatrixOrder.Prepend);
                }
        
                public void Rotate (float angle, MatrixOrder order)
                {
			Status status = GDIPlus.GdipRotateMatrix (nativeMatrix, angle, order);
			GDIPlus.CheckStatus (status);
                }
        
                public void RotateAt (float angle, PointF point)
                {
                        RotateAt (angle, point, MatrixOrder.Prepend);
                }
        
                public void RotateAt (float angle, PointF point, MatrixOrder order)
                {
                        angle *= (float) (Math.PI / 180.0);  // degrees to radians
                        float cos = (float) Math.Cos (angle);
                        float sin = (float) Math.Sin (angle);
                        float e4 = -point.X * cos + point.Y * sin + point.X;
                        float e5 = -point.X * sin - point.Y * cos + point.Y;
                        float[] m = this.Elements;

			Status status;

                        if (order == MatrixOrder.Prepend)
				status = GDIPlus.GdipSetMatrixElements (nativeMatrix,
								cos * m[0] + sin * m[2],
								cos * m[1] + sin * m[3],
								-sin * m[0] + cos * m[2],
								-sin * m[1] + cos * m[3],
								e4 * m[0] + e5 * m[2] + m[4],
								e4 * m[1] + e5 * m[3] + m[5]);
                        else
				status = GDIPlus.GdipSetMatrixElements (nativeMatrix,
								m[0] * cos + m[1] * -sin,
								m[0] * sin + m[1] * cos,
								m[2] * cos + m[3] * -sin,
								m[2] * sin + m[3] * cos,
								m[4] * cos + m[5] * -sin + e4,
								m[4] * sin + m[5] * cos + e5);
			GDIPlus.CheckStatus (status);
                }
        
                public void Scale (float scaleX, float scaleY)
                {
                        Scale (scaleX, scaleY, MatrixOrder.Prepend);
                }
        
                public void Scale (float scaleX, float scaleY, MatrixOrder order)
                {
			Status status = GDIPlus.GdipScaleMatrix (nativeMatrix, scaleX, scaleY, order);
			GDIPlus.CheckStatus (status);
                }
        
                public void Shear (float shearX, float shearY)
                {
                        Shear (shearX, shearY, MatrixOrder.Prepend);
                }
        
                public void Shear (float shearX, float shearY, MatrixOrder order)
                {
			Status status = GDIPlus.GdipShearMatrix (nativeMatrix, shearX, shearY, order);
			GDIPlus.CheckStatus (status);
                }
        
                public void TransformPoints (Point[] pts)
                {
			Status status = GDIPlus.GdipTransformMatrixPointsI (nativeMatrix, pts, pts.Length);
			GDIPlus.CheckStatus (status);
                }
        
                public void TransformPoints (PointF[] pts)
                {
			Status status = GDIPlus.GdipTransformMatrixPoints (nativeMatrix, pts, pts.Length);
			GDIPlus.CheckStatus (status);
                }
        
                public void TransformVectors (Point[] pts)
                {
			Status status = GDIPlus.GdipVectorTransformMatrixPointsI (nativeMatrix, pts, pts.Length);
			GDIPlus.CheckStatus (status);
                }
        
                public void TransformVectors (PointF[] pts)
                {
			Status status = GDIPlus.GdipVectorTransformMatrixPoints (nativeMatrix, pts, pts.Length);
			GDIPlus.CheckStatus (status);
                }
        
                public void Translate (float offsetX, float offsetY)
                {
                        Translate (offsetX, offsetY, MatrixOrder.Prepend);
                }
        
                public void Translate (float offsetX, float offsetY, MatrixOrder order)
                {
			Status status = GDIPlus.GdipTranslateMatrix (nativeMatrix, offsetX, offsetY, order);
			GDIPlus.CheckStatus (status);
                }
        
                public void VectorTransformPoints (Point[] pts)
                {
                        TransformVectors (pts);
                }
                
                internal IntPtr NativeObject
                {
			get{
				return nativeMatrix;
			}
			set	{
				nativeMatrix = value;
			}
		}
        }
}

File: mcs/class/System.Drawing/gdiplus/matrix.c
/**
 * matrix.c
 *
 * Author: Duncan Mak (duncan@ximian.com)
 * 
 * Copyright (C) Novell, Inc. 2003.
 *
 **/

#include <math.h>
#include "gdip.h"

GpStatus 
GdipCreateMatrix (GpMatrix **matrix)
{
        *matrix = cairo_matrix_create ();
        cairo_matrix_set_affine (*matrix, 1, 0, 0, 1, 0, 0);
        return Ok;
}

GpStatus
GdipCreateMatrix2 (float m11, float m12, float m21, float m22, float dx, float dy, GpMatrix **matrix)
{
        *matrix = cairo_matrix_create ();

        return gdip_get_status (
                cairo_matrix_set_affine (
                        *matrix, m11, m12, m21, m22, dx, dy));
}

GpStatus
GdipCreateMatrix3 (const GpRectF *rect, const GpPointF *dstplg, GpMatrix **matrix)
{
        *matrix = cairo_matrix_create ();
        double m11 = rect->left;
        double m12 = rect->top;
        double m21 = (rect->right) - (rect->left); /* width */
        double m22 = (rect->bottom) - (rect->top); /* height */
        double dx = dstplg->X;
        double dy = dstplg->Y;

        Status s = cairo_matrix_set_affine (
                        *matrix, m11, m12, m21, m22, dx, dy);

        return gdip_get_status (s);
}

GpStatus
GdipCreateMatrix3I (const GpRect *rect, const GpPoint *dstplg, GpMatrix **matrix)
{
        *matrix = cairo_matrix_create ();
        double m11 = rect->left;
        double m12 = rect->top;
        double m21 = (rect->right) - (rect->left); /* width */
        double m22 = (rect->bottom) - (rect->top); /* height */
        double dx = dstplg->X;
        double dy = dstplg->Y;
        
        Status s = cairo_matrix_set_affine (
                        *matrix, m11, m12, m21, m22, dx, dy);

        return gdip_get_status (s);
}

GpStatus
GdipCloneMatrix (GpMatrix *matrix, GpMatrix **cloneMatrix)
{
        *cloneMatrix = cairo_matrix_create ();
        
        return gdip_get_status (
                cairo_matrix_copy (*cloneMatrix, matrix));
}

GpStatus
GdipDeleteMatrix (GpMatrix *matrix)
{
        if (matrix)
                cairo_matrix_destroy (matrix);
        return Ok;
}

GpStatus
GdipSetMatrixElements (GpMatrix *matrix, float m11, float m12, float m21, float m22, float dx, float dy)
{
        return gdip_get_status (
                cairo_matrix_set_affine (matrix, m11, m12, m21, m22, dx, dy));
}

GpStatus 
GdipGetMatrixElements (GpMatrix *matrix, float *matrixOut)
{
        double a, b, c, d, tx, ty;
        
        cairo_matrix_get_affine (matrix, &a, &b, &c, &d, &tx, &ty);
        
        matrixOut[0] = (float) a;
        matrixOut[1] = (float) b;
        matrixOut[2] = (float) c;
        matrixOut[3] = (float) d;
        matrixOut[4] = (float) tx;
        matrixOut[5] = (float) ty;

        return Ok;
}

GpStatus
GdipMultiplyMatrix (GpMatrix *matrix, GpMatrix *matrix2, GpMatrixOrder order)
{
        cairo_status_t status;

        if (order == MatrixOrderAppend)
                status = cairo_matrix_multiply (matrix, matrix, matrix2);

        else if (order == MatrixOrderPrepend)
                status = cairo_matrix_multiply (matrix, matrix2, matrix);
        
        else
                return GenericError;

        return gdip_get_status (status);
}

static GpMatrix *
set_translate (float offsetX, float offsetY)
{
        GpMatrix *matrix = cairo_matrix_create ();
        cairo_matrix_set_affine (matrix, 1, 0, 0, 1, offsetX, offsetY);
        return matrix;
}

GpStatus
GdipTranslateMatrix (GpMatrix *matrix, float offsetX, float offsetY, GpMatrixOrder order)
{
        GpMatrix *tmp = set_translate (offsetX, offsetY);
        GpStatus s = GdipMultiplyMatrix (matrix, tmp, order);
        GdipDeleteMatrix (tmp);

        return s;
}

static GpMatrix *
set_scale (float scaleX, float scaleY)
{
        GpMatrix *matrix = cairo_matrix_create ();
        cairo_matrix_set_affine (matrix, scaleX, 0, 0, scaleY, 0, 0);
        return matrix;
}

GpStatus
GdipScaleMatrix (GpMatrix *matrix, float scaleX, float scaleY, GpMatrixOrder order)
{
        GpMatrix *tmp = set_scale (scaleX, scaleY);
        GpStatus s = GdipMultiplyMatrix (matrix, tmp, order);
        GdipDeleteMatrix (tmp);

        return s;
}

static GpMatrix *
set_rotate (float angle)
{
        float rad = angle * DEGTORAD;
        GpMatrix *matrix = cairo_matrix_create ();
        cairo_matrix_set_affine (matrix, cos (rad), sin (rad), -sin (rad), cos (rad), 0, 0);

        return matrix;
}

GpStatus
GdipRotateMatrix (GpMatrix *matrix, float angle, GpMatrixOrder order)
{
        GpMatrix *tmp = set_rotate (angle);
        GpStatus s = GdipMultiplyMatrix (matrix, tmp, order);
        GdipDeleteMatrix (tmp);

        return s;
}

static GpMatrix *
set_shear (float shearX, float shearY)
{
        GpMatrix *matrix = cairo_matrix_create ();
        cairo_matrix_set_affine (matrix, 1, shearX, shearY, 1, 0, 0);
        return matrix;
}

GpStatus
GdipShearMatrix (GpMatrix *matrix, float shearX, float shearY, GpMatrixOrder order)
{
        GpMatrix *tmp = set_shear (shearX, shearY);
        GpStatus s = GdipMultiplyMatrix (matrix, tmp, order);
        GdipDeleteMatrix (tmp);

        return s;
}

GpStatus
GdipInvertMatrix (GpMatrix *matrix)
{
        return gdip_get_status (
                cairo_matrix_invert (matrix));
}

GpStatus
GdipTransformMatrixPoints (GpMatrix *matrix, GpPointF *pts, int count)
{
        int i;
        cairo_status_t status;
        
        for (i = 0; i < count; i++, pts++) {
                double x = pts->X;
                double y = pts->Y;
                status = cairo_matrix_transform_point (matrix, &x, &y);
                if (status != CAIRO_STATUS_SUCCESS)
                        return gdip_get_status (status);

                pts->X = (float) x;
                pts->Y = (float) y;
        }

        return Ok;
}

GpStatus
GdipTransformMatrixPointsI (GpMatrix *matrix, GpPoint *pts, int count)
{
        int i;
        cairo_status_t status;
        
        for (i = 0; i < count; i++, pts++) {
                double x = pts->X;
                double y = pts->Y;
                status = cairo_matrix_transform_point (matrix, &x, &y);
                if (status != CAIRO_STATUS_SUCCESS)
                        return gdip_get_status (status);
                pts->X = (int) x;
                pts->Y = (int) y;
        }

        return Ok;
}

GpStatus
GdipVectorTransformMatrixPoints (GpMatrix *matrix, GpPointF *pts, int count)
{
        int i;
        cairo_status_t status;
        
        for (i = 0; i < count; i++, pts++) {
                double x = pts->X;
                double y = pts->Y;
                status = cairo_matrix_transform_distance (matrix, &x, &y);
                if (status != CAIRO_STATUS_SUCCESS)
                        return gdip_get_status (status);
                pts->X = (float) x;
                pts->Y = (float) y;
        }

        return Ok;
}

GpStatus
GdipVectorTransformMatrixPointsI (GpMatrix *matrix, GpPoint *pts, int count)
{
        int i;
        cairo_status_t status;
        
        for (i = 0; i < count; i++, pts++) {
                double x = pts->X;
                double y = pts->Y;
                status = cairo_matrix_transform_distance (matrix, &x, &y);
                if (status != CAIRO_STATUS_SUCCESS)
                        return gdip_get_status (status);

                pts->X = (int) x;
                pts->Y = (int) y;
        }

        return Ok;
}

GpStatus 
GdipIsMatrixInvertible (GpMatrix *matrix, int *result)
{
        cairo_status_t status = cairo_matrix_invert (matrix);

        if (status == CAIRO_STATUS_INVALID_MATRIX)
                *result = 0;

        *result = 1;
        return Ok;
}

static int
matrix_equals (GpMatrix *x, GpMatrix *y)
{
        double ax, bx, cx, dx, ex, fx;
        double ay, by, cy, dy, ey, fy;

        cairo_matrix_get_affine (x, &ax, &bx, &cx, &dx, &ex, &fx);
        cairo_matrix_get_affine (y, &ay, &by, &cy, &dy, &ey, &fy);

        if ((ax != ay) || (bx != by) || (cx != cy) ||
            (dx != dy) || (ex != ey) || (fx != fy))
                return 0;

        return 1;
}

GpStatus
GdipIsMatrixIdentity (GpMatrix *matrix, int *result)
{
        GpMatrix *identity = cairo_matrix_create ();
        cairo_matrix_set_identity (identity);

        Status s = GdipIsMatrixEqual (matrix, identity, result);
        GdipDeleteMatrix (identity);

        return s;
}

GpStatus
GdipIsMatrixEqual (GpMatrix *matrix, GpMatrix *matrix2, int *result)
{
        *result = matrix_equals (matrix, matrix2);
        return Ok;
}
